052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,147 @@
+
+There is more documentation within the [Merge Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) documentation in which it defines how we can change CRDs to merge how we want them to (StrimziPodSet, for example). For reference, there are two main kinds: `atomic` and `map`/`set`/`granular`; with `atomic` set, any change replaces the whole field/list and is recursive downwards; with the others the behaviour is more fine-tuned and is described on the above link in a table.
+
+### Code changes
+
+We need to:
+* Modify the PATCH/CREATE commands to server-side apply
+* Remove the GET commands
+* Change the ResourceDiff (and other similar functions) to only look at `strimzi-cluster-operator` owned resources (within `metadata.managedFields`)
+* Update tests
+
+---
+
+### **Modify the PATCH/CREATE commands to server-side apply**
+Modify [this line](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#LL263C88-L263C102).
+From:
+
+```
+....patch(PatchContext.of(PatchType.JSON), desired);
+```
+
+To:
+
+<!-- Note: change the below examples to serverSideApply once we know how -->
+```
+....patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withForce(true).build(), desired);
+```
+
+Also, the `StrimziPodSetOperator` patches this method [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StrimziPodSetOperator.java#L65), this would need updating too. 
+It would also need confirming that all resource creates/updates go through `serverSideApply`, not something that should be done in this proposal.
+
+And [Kubernetes](https://kubernetes.io/docs/reference/using-api/server-side-apply/#:~:text=through%20declarative%20configurations.-,Clients%20can%20create,-and%20modify%20their)/[Fabric8io](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#server-side-apply:~:text=For%20any%20create%20or%20update) recommend it is used on CREATE also:
+[From](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#L272):
+
+```
+ReconcileResult<T> result = ReconcileResult.created(operation().inNamespace(namespace).resource(desired).create());
+```
+
+To:
+
+```
+ReconcileResult<T> result = ReconcileResult.patched(operation().inNamespace(namespace).withName(name).patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withForce(true).build(), desired);
```

Comment:
Something that was not covered so far and should be probably added as well ... it looks like by default, Fabric8 will set the field manager to `fabric8`. I guess we should set it to some custom value? E.g. `strimzi-cluster-operator`? It seems to be possible only in the patch context I guess?

Reply:
Actually, I'm not sure. In an existing cluster I have, it seems to be set to `strimzi-cluster-operator` as the field manager. Not sure if that was deduced from the service account? If yes, why running the client locally does use `fabric8`? I think this should be figured out. For the server side apply to work properly, it should ideally use a single field manager and it should be the same field manager all the time otherwise there will be a mess in the patching of any maps. So I think this should be discussed and included in this proposal as well.

Reply:
Yes I noticed the same thing when I looked at Strimzi field owners recently, [kubernetes state that](https://kubernetes.io/docs/reference/using-api/server-side-apply/#:~:text=its%20default%20is%20computed%20from%20the%20user%2Dagent.) "its default is computed from the user-agent."

So I guess setting it every time for every call we make to Kubernetes will make it consistent.

Reply:
This is something that would need the code to change to discover, or we can try and find an example operator setting the `fieldManager` and install it locally and check it out

Reply:
Looks like [this line](https://github.com/strimzi/strimzi-kafka-operator/blob/4a5bec44d8e2c4dd55bfd5d5764afd6639afe230/cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java#L89) is why some resources have `strimzi-cluster-operator`